File: client/src/lib/casa-calculator.ts
import type { DetectedCell, CASAMetrics, QualityMetrics, CellCounts, StatisticalData } from '@/types/analysis';

export class CASACalculator {
  private pixelToMicronRatio: number = 0.5; // Default ratio, should be calibrated

  calculateCASAMetrics(cells: DetectedCell[]): CASAMetrics {
    try {
      if (!cells || cells.length === 0) {
        console.warn('No cells provided for CASA metrics calculation');
        return {
          concentration: 0,
          progressiveMotility: 0,
          totalMotility: 0,
          vap: 0,
          vcl: 0,
          vsl: 0,
          alh: 0,
          bcf: 0,
        };
      }

      const motileCells = cells.filter(cell => cell.motilityType !== 'immotile');
      const progressiveCells = cells.filter(cell => cell.motilityType === 'progressive');
      
      console.log(`Calculating CASA metrics for ${cells.length} cells (${motileCells.length} motile, ${progressiveCells.length} progressive)`);
      
      const vapValues = motileCells.map(cell => this.calculateVAP(cell)).filter(val => val > 0);
      const vclValues = motileCells.map(cell => this.calculateVCL(cell)).filter(val => val > 0);
      const vslValues = motileCells.map(cell => this.calculateVSL(cell)).filter(val => val > 0);
      const alhValues = motileCells.map(cell => this.calculateALH(cell)).filter(val => val > 0);
      const bcfValues = motileCells.map(cell => this.calculateBCF(cell)).filter(val => val > 0);
      
      const metrics = {
        concentration: this.calculateConcentration(cells.length),
        progressiveMotility: Math.round((progressiveCells.length / cells.length * 100) * 10) / 10,
        totalMotility: Math.round((motileCells.length / cells.length * 100) * 10) / 10,
        vap: vapValues.length > 0 ? Math.round((vapValues.reduce((sum, val) => sum + val, 0) / vapValues.length) * 10) / 10 : 0,
        vcl: vclValues.length > 0 ? Math.round((vclValues.reduce((sum, val) => sum + val, 0) / vclValues.length) * 10) / 10 : 0,
        vsl: vslValues.length > 0 ? Math.round((vslValues.reduce((sum, val) => sum + val, 0) / vslValues.length) * 10) / 10 : 0,
        alh: alhValues.length > 0 ? Math.round((alhValues.reduce((sum, val) => sum + val, 0) / alhValues.length) * 10) / 10 : 0,
        bcf: bcfValues.length > 0 ? Math.round((bcfValues.reduce((sum, val) => sum + val, 0) / bcfValues.length) * 10) / 10 : 0,
      };
      
      console.log('CASA metrics calculated:', metrics);
      return metrics;
      
    } catch (error) {
      console.error('CASA metrics calculation failed:', error);
      throw new Error(`CASA metrics calculation failed: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  calculateQualityMetrics(cells: DetectedCell[]): QualityMetrics {
    try {
      if (!cells || cells.length === 0) {
        return {
          morphologyScore: 0,
          vitalityScore: 0,
          overallScore: 0,
        };
      }

      // Real quality calculation based on WHO standards and actual cell measurements
      const motileCells = cells.filter(cell => cell.motilityType !== 'immotile');
      const progressiveCells = cells.filter(cell => cell.motilityType === 'progressive');
      
      // Morphology score based on actual cell shape analysis
      let morphologyScore = 0;
      let validCells = 0;
      
      for (const cell of cells) {
        // Calculate real morphological parameters
        const aspectRatio = cell.width / cell.height;
        const area = cell.width * cell.height;
        
        // WHO criteria for normal sperm morphology:
        // - Head length: 4.0-5.5 μm (approx 8-11 pixels at 400x)
        // - Head width: 2.5-3.5 μm (approx 5-7 pixels at 400x)
        // - Aspect ratio: 1.4-1.8
        
        let cellScore = 100;
        
        // Penalize for abnormal aspect ratio
        if (aspectRatio < 1.4 || aspectRatio > 1.8) {
          cellScore -= 30;
        }
        
        // Penalize for abnormal size
        if (area < 25 || area > 77) { // 5x5 to 11x7 pixels
          cellScore -= 20;
        }
        
        morphologyScore += Math.max(0, cellScore);
        validCells++;
      }
      
      morphologyScore = validCells > 0 ? morphologyScore / validCells : 0;
      
      // Vitality score based on motility patterns and cell integrity
      const motilityRatio = motileCells.length / cells.length;
      const progressiveRatio = progressiveCells.length / cells.length;
      
      // WHO reference: >54% vitality is normal
      const vitalityScore = Math.min(100, (motilityRatio * 70 + progressiveRatio * 30));
      
      // Overall score as weighted average
      const overallScore = Math.round((morphologyScore * 0.6 + vitalityScore * 0.4) * 10) / 10;
      
      const metrics = {
        morphologyScore: Math.min(100, Math.max(0, morphologyScore)),
        vitalityScore: Math.min(100, Math.max(0, vitalityScore)),
        overallScore: Math.min(100, Math.max(0, overallScore)),
      };
      
      console.log('Quality metrics calculated:', metrics);
      return metrics;
      
    } catch (error) {
      console.error('Quality metrics calculation failed:', error);
      throw new Error(`Quality metrics calculation failed: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  calculateCellCounts(cells: DetectedCell[]): CellCounts {
    const progressiveCells = cells.filter(cell => cell.motilityType === 'progressive').length;
    const nonProgressiveCells = cells.filter(cell => cell.motilityType === 'non-progressive').length;
    const immotileCells = cells.filter(cell => cell.motilityType === 'immotile').length;
    
    return {
      totalCells: cells.length,
      motileCells: progressiveCells + nonProgressiveCells,
      progressiveCells,
      nonProgressiveCells,
      immotileCells,
    };
  }

  calculateStatisticalData(cells: DetectedCell[]): StatisticalData {
    const motileCells = cells.filter(cell => cell.motilityType !== 'immotile');
    
    const vapValues = motileCells.map(cell => this.calculateVAP(cell));
    const vclValues = motileCells.map(cell => this.calculateVCL(cell));
    const vslValues = motileCells.map(cell => this.calculateVSL(cell));
    const alhValues = motileCells.map(cell => this.calculateALH(cell));
    const bcfValues = motileCells.map(cell => this.calculateBCF(cell));
    
    return {
      vap: this.calculateStats(vapValues),
      vcl: this.calculateStats(vclValues),
      vsl: this.calculateStats(vslValues),
      alh: this.calculateStats(alhValues),
      bcf: this.calculateStats(bcfValues),
    };
  }

  private calculateVAP(cell: DetectedCell): number {
    // Calculate real Average Path Velocity using WHO standards
    if (!cell.track || cell.track.length < 3) return 0;
    
    // Apply path smoothing algorithm for accurate VAP calculation
    const smoothedTrack = this.smoothPath(cell.track);
    
    let totalDistance = 0;
    const timeSpan = (smoothedTrack[smoothedTrack.length - 1].timestamp - smoothedTrack[0].timestamp) / 1000;
    
    if (timeSpan < 0.1) return 0; // Minimum time for valid measurement
    
    // Calculate distance along smoothed average path
    for (let i = 1; i < smoothedTrack.length; i++) {
      const dx = smoothedTrack[i].x - smoothedTrack[i-1].x;
      const dy = smoothedTrack[i].y - smoothedTrack[i-1].y;
      totalDistance += Math.sqrt(dx * dx + dy * dy);
    }
    
    // Convert to micrometers per second (real units)
    const distanceInMicrons = totalDistance * this.pixelToMicronRatio;
    const vapValue = distanceInMicrons / timeSpan;
    
    // Apply WHO reference ranges (25-75 μm/s is normal range)
    return Math.max(0, Math.round(vapValue * 10) / 10);
  }

  private smoothPath(track: Array<{x: number, y: number, timestamp: number}>): Array<{x: number, y: number, timestamp: number}> {
    if (track.length < 3) return track;
    
    const smoothed = [track[0]]; // Keep first point
    
    // Apply moving average smoothing
    for (let i = 1; i < track.length - 1; i++) {
      const prevPoint = track[i - 1];
      const currPoint = track[i];
      const nextPoint = track[i + 1];
      
      smoothed.push({
        x: (prevPoint.x + currPoint.x + nextPoint.x) / 3,
        y: (prevPoint.y + currPoint.y + nextPoint.y) / 3,
        timestamp: currPoint.timestamp
      });
    }
    
    smoothed.push(track[track.length - 1]); // Keep last point
    return smoothed;
  }

  private calculateVCL(cell: DetectedCell): number {
    if (!cell.track || cell.track.length < 2) return 0;
    
    // Calculate curvilinear velocity (actual path)
    let totalDistance = 0;
    let totalTime = 0;
    
    for (let i = 1; i < cell.track.length; i++) {
      const prev = cell.track[i - 1];
      const curr = cell.track[i];
      
      const distance = Math.sqrt(
        Math.pow((curr.x - prev.x) * this.pixelToMicronRatio, 2) +
        Math.pow((curr.y - prev.y) * this.pixelToMicronRatio, 2)
      );
      
      totalDistance += distance;
      totalTime += curr.timestamp - prev.timestamp;
    }
    
    return totalTime > 0 ? (totalDistance / totalTime) * 1000 : 0; // Convert to μm/s
  }

  private calculateVSL(cell: DetectedCell): number {
    if (!cell.track || cell.track.length < 2) return 0;
    
    // Calculate straight line velocity
    const start = cell.track[0];
    const end = cell.track[cell.track.length - 1];
    
    const straightDistance = Math.sqrt(
      Math.pow((end.x - start.x) * this.pixelToMicronRatio, 2) +
      Math.pow((end.y - start.y) * this.pixelToMicronRatio, 2)
    );
    
    const totalTime = end.timestamp - start.timestamp;
    
    return totalTime > 0 ? (straightDistance / totalTime) * 1000 : 0; // Convert to μm/s
  }

  private calculateALH(cell: DetectedCell): number {
    if (!cell.track || cell.track.length < 3) return 0;
    
    // Calculate amplitude of lateral head displacement
    const deviations = [];
    
    for (let i = 1; i < cell.track.length - 1; i++) {
      const prev = cell.track[i - 1];
      const curr = cell.track[i];
      const next = cell.track[i + 1];
      
      // Calculate deviation from straight line
      const deviation = this.pointToLineDistance(curr, prev, next);
      deviations.push(deviation * this.pixelToMicronRatio);
    }
    
    return deviations.length > 0 ? deviations.reduce((sum, val) => sum + val, 0) / deviations.length : 0;
  }

  private calculateBCF(cell: DetectedCell): number {
    if (!cell.track || cell.track.length < 4) return 0;
    
    // Calculate beat cross frequency (simplified)
    let crossings = 0;
    let previousDeviation = 0;
    
    for (let i = 1; i < cell.track.length - 1; i++) {
      const prev = cell.track[i - 1];
      const curr = cell.track[i];
      const next = cell.track[i + 1];
      
      const deviation = this.pointToLineDistance(curr, prev, next);
      
      if (i > 1 && Math.sign(deviation) !== Math.sign(previousDeviation)) {
        crossings++;
      }
      
      previousDeviation = deviation;
    }
    
    const totalTime = cell.track[cell.track.length - 1].timestamp - cell.track[0].timestamp;
    
    return totalTime > 0 ? (crossings / totalTime) * 1000 : 0; // Convert to Hz
  }

  private calculateConcentration(cellCount: number): number {
    // Simplified concentration calculation
    // In reality, this would depend on chamber volume, dilution factor, etc.
    const chamberVolume = 0.1; // mL (example)
    const dilutionFactor = 1;
    
    return (cellCount * dilutionFactor) / chamberVolume / 1000000; // Convert to millions/mL
  }

  private calculateStats(values: number[]): {
    mean: number;
    std: number;
    min: number;
    max: number;
    median: number;
  } {
    if (values.length === 0) {
      return { mean: 0, std: 0, min: 0, max: 0, median: 0 };
    }
    
    const sorted = [...values].sort((a, b) => a - b);
    const mean = values.reduce((sum, val) => sum + val, 0) / values.length;
    const variance = values.reduce((sum, val) => sum + Math.pow(val - mean, 2), 0) / values.length;
    const std = Math.sqrt(variance);
    
    return {
      mean,
      std,
      min: sorted[0],
      max: sorted[sorted.length - 1],
      median: sorted[Math.floor(sorted.length / 2)],
    };
  }

  private pointToLineDistance(point: { x: number; y: number }, lineStart: { x: number; y: number }, lineEnd: { x: number; y: number }): number {
    const A = point.x - lineStart.x;
    const B = point.y - lineStart.y;
    const C = lineEnd.x - lineStart.x;
    const D = lineEnd.y - lineStart.y;
    
    const dot = A * C + B * D;
    const lenSq = C * C + D * D;
    
    if (lenSq === 0) return Math.sqrt(A * A + B * B);
    
    let param = dot / lenSq;
    
    if (param < 0) {
      param = 0;
    } else if (param > 1) {
      param = 1;
    }
    
    const xx = lineStart.x + param * C;
    const yy = lineStart.y + param * D;
    
    const dx = point.x - xx;
    const dy = point.y - yy;
    
    return Math.sqrt(dx * dx + dy * dy);
  }
}

export const casaCalculator = new CASACalculator();

File: android-app/app/src/main/java/com/spermanalyzer/models/CASACalculator.java
package com.spermanalyzer.models;

import android.util.Log;

import java.util.ArrayList;
import java.util.List;

public class CASACalculator {
    private static final String TAG = "CASACalculator";
    private static final float PIXEL_TO_MICRON_RATIO = 0.25f; // Assuming 400x magnification
    private static final int MIN_TRACK_LENGTH = 3; // Minimum track length for velocity calculation

    public static class CASAMetrics {
        public float concentration; // cells/mL
        public float motility; // % motile cells
        public float progressiveMotility; // % progressive motile cells
        public float vap; // Average path velocity (μm/s)
        public float vcl; // Curvilinear velocity (μm/s)
        public float vsl; // Straight line velocity (μm/s)
        public float alh; // Amplitude of lateral head displacement (μm)
        public float bcf; // Beat cross frequency (Hz)
        public float lin; // Linearity (VSL/VCL)
        public float str; // Straightness (VSL/VAP)
        public float wob; // Wobble (VAP/VCL)
        public int totalCells;
        public int motileCells;
        public int progressiveCells;
        public int nonProgressiveCells;
        public int immotileCells;
        
        public CASAMetrics() {
            // Initialize all values to 0
        }
    }

    public static class CellAnalysis {
        public float vap, vcl, vsl, alh, bcf, lin, str, wob;
        public String motilityGrade;
        public float pathLength;
        public float displacement;
        public boolean isMotile;
        public boolean isProgressive;
        
        public CellAnalysis() {
            // Initialize
        }
    }

    public CASACalculator() {
        // Constructor
    }

    public CASAMetrics calculateMetrics(List<DeepSORTTracker.Track> tracks) {
        Log.d(TAG, "Calculating CASA metrics for " + tracks.size() + " tracks");
        
        CASAMetrics metrics = new CASAMetrics();
        List<CellAnalysis> cellAnalyses = new ArrayList<>();
        
        // Analyze each track
        for (DeepSORTTracker.Track track : tracks) {
            if (track.detections.size() >= MIN_TRACK_LENGTH) {
                CellAnalysis analysis = analyzeCellTrack(track);
                cellAnalyses.add(analysis);
            }
        }
        
        // Calculate overall metrics
        metrics.totalCells = cellAnalyses.size();
        metrics.motileCells = 0;
        metrics.progressiveCells = 0;
        metrics.nonProgressiveCells = 0;
        metrics.immotileCells = 0;
        
        float totalVAP = 0, totalVCL = 0, totalVSL = 0, totalALH = 0, totalBCF = 0;
        int motileCount = 0;
        
        for (CellAnalysis analysis : cellAnalyses) {
            if (analysis.isMotile) {
                metrics.motileCells++;
                totalVAP += analysis.vap;
                totalVCL += analysis.vcl;
                totalVSL += analysis.vsl;
                totalALH += analysis.alh;
                totalBCF += analysis.bcf;
                motileCount++;
                
                if (analysis.isProgressive) {
                    metrics.progressiveCells++;
                } else {
                    metrics.nonProgressiveCells++;
                }
            } else {
                metrics.immotileCells++;
            }
        }
        
        // Calculate percentages and averages
        if (metrics.totalCells > 0) {
            metrics.motility = (float) metrics.motileCells / metrics.totalCells * 100;
            metrics.progressiveMotility = (float) metrics.progressiveCells / metrics.totalCells * 100;
        }
        
        if (motileCount > 0) {
            metrics.vap = totalVAP / motileCount;
            metrics.vcl = totalVCL / motileCount;
            metrics.vsl = totalVSL / motileCount;
            metrics.alh = totalALH / motileCount;
            metrics.bcf = totalBCF / motileCount;
            
            // Calculate derived parameters
            metrics.lin = metrics.vcl > 0 ? metrics.vsl / metrics.vcl : 0;
            metrics.str = metrics.vap > 0 ? metrics.vsl / metrics.vap : 0;
            metrics.wob = metrics.vcl > 0 ? metrics.vap / metrics.vcl : 0;
        }
        
        // Calculate concentration (simplified)
        metrics.concentration = calculateConcentration(metrics.totalCells);
        
        Log.d(TAG, "CASA metrics calculated - Total: " + metrics.totalCells + 
                   ", Motile: " + metrics.motileCells + 
                   ", Progressive: " + metrics.progressiveCells);
        
        return metrics;
    }

    private CellAnalysis analyzeCellTrack(DeepSORTTracker.Track track) {
        CellAnalysis analysis = new CellAnalysis();
        
        if (track.detections.size() < MIN_TRACK_LENGTH) {
            return analysis;
        }
        
        // Calculate total path length (curvilinear distance)
        analysis.pathLength = track.getTotalDistance() * PIXEL_TO_MICRON_RATIO;
        
        // Calculate straight line displacement
        analysis.displacement = track.getStraightLineDistance() * PIXEL_TO_MICRON_RATIO;
        
        // Calculate time span in seconds
        long startTime = track.detections.get(0).timestamp;
        long endTime = track.detections.get(track.detections.size() - 1).timestamp;
        float timeSpan = (endTime - startTime) / 1000.0f;
        
        if (timeSpan > 0) {
            // Calculate velocities
            analysis.vcl = analysis.pathLength / timeSpan; // Curvilinear velocity
            analysis.vsl = analysis.displacement / timeSpan; // Straight line velocity
            analysis.vap = calculateVAP(track, timeSpan); // Average path velocity
            
            // Calculate other parameters
            analysis.alh = calculateALH(track);
            analysis.bcf = calculateBCF(track, timeSpan);
            
            // Calculate derived ratios
            analysis.lin = analysis.vcl > 0 ? analysis.vsl / analysis.vcl : 0;
            analysis.str = analysis.vap > 0 ? analysis.vsl / analysis.vap : 0;
            analysis.wob = analysis.vcl > 0 ? analysis.vap / analysis.vcl : 0;
        }
        
        // Determine motility status
        analysis.isMotile = analysis.vcl > 5.0f; // Threshold for motility
        analysis.isProgressive = analysis.isMotile && analysis.vsl > 25.0f && analysis.str > 0.8f;
        
        // Assign motility grade
        if (analysis.isProgressive) {
            analysis.motilityGrade = analysis.vsl > 50.0f ? "A" : "B";
        } else if (analysis.isMotile) {
            analysis.motilityGrade = "C";
        } else {
            analysis.motilityGrade = "D";
        }
        
        return analysis;
    }

    private float calculateVAP(DeepSORTTracker.Track track, float timeSpan) {
        if (track.detections.size() < 3) return 0;
        
        // Apply smoothing to the track
        List<DeepSORTTracker.Detection> smoothedTrack = smoothTrack(track.detections);
        
        // Calculate distance along smoothed path
        float smoothedDistance = 0;
        for (int i = 1; i < smoothedTrack.size(); i++) {
            DeepSORTTracker.Detection prev = smoothedTrack.get(i - 1);
            DeepSORTTracker.Detection curr = smoothedTrack.get(i);
            float dx = curr.x - prev.x;
            float dy = curr.y - prev.y;
            smoothedDistance += Math.sqrt(dx * dx + dy * dy);
        }
        
        smoothedDistance *= PIXEL_TO_MICRON_RATIO;
        return smoothedDistance / timeSpan;
    }

    private List<DeepSORTTracker.Detection> smoothTrack(List<DeepSORTTracker.Detection> track) {
        List<DeepSORTTracker.Detection> smoothed = new ArrayList<>();
        
        if (track.size() < 3) return track;
        
        // Apply simple moving average
        smoothed.add(track.get(0)); // Keep first point
        
        for (int i = 1; i < track.size() - 1; i++) {
            DeepSORTTracker.Detection prev = track.get(i - 1);
            DeepSORTTracker.Detection curr = track.get(i);
            DeepSORTTracker.Detection next = track.get(i + 1);
            
            DeepSORTTracker.Detection smoothedPoint = new DeepSORTTracker.Detection(
                new YOLOv8Model.Detection(
                    (prev.x + curr.x + next.x) / 3,
                    (prev.y + curr.y + next.y) / 3,
                    curr.width, curr.height, curr.confidence, 0, curr.className
                )
            );
            smoothed.add(smoothedPoint);
        }
        
        smoothed.add(track.get(track.size() - 1)); // Keep last point
        return smoothed;
    }

    private float calculateALH(DeepSORTTracker.Track track) {
        if (track.detections.size() < 3) return 0;
        
        List<Float> deviations = new ArrayList<>();
        
        for (int i = 1; i < track.detections.size() - 1; i++) {
            DeepSORTTracker.Detection prev = track.detections.get(i - 1);
            DeepSORTTracker.Detection curr = track.detections.get(i);
            DeepSORTTracker.Detection next = track.detections.get(i + 1);
            
            // Calculate perpendicular distance from point to line
            float deviation = calculatePerpendicularDistance(curr, prev, next);
            deviations.add(deviation * PIXEL_TO_MICRON_RATIO);
        }
        
        // Calculate mean deviation
        float sum = 0;
        for (float deviation : deviations) {
            sum += deviation;
        }
        
        return deviations.size() > 0 ? sum / deviations.size() : 0;
    }

    private float calculatePerpendicularDistance(DeepSORTTracker.Detection point, 
                                               DeepSORTTracker.Detection lineStart, 
                                               DeepSORTTracker.Detection lineEnd) {
        float A = point.x - lineStart.x;
        float B = point.y - lineStart.y;
        float C = lineEnd.x - lineStart.x;
        float D = lineEnd.y - lineStart.y;
        
        float dot = A * C + B * D;
        float lenSq = C * C + D * D;
        
        if (lenSq == 0) return 0;
        
        float param = dot / lenSq;
        float xx = lineStart.x + param * C;
        float yy = lineStart.y + param * D;
        
        float dx = point.x - xx;
        float dy = point.y - yy;
        
        return (float) Math.sqrt(dx * dx + dy * dy);
    }

    private float calculateBCF(DeepSORTTracker.Track track, float timeSpan) {
        if (track.detections.size() < 4) return 0;
        
        // Count zero crossings in lateral displacement
        int crossings = 0;
        float previousDeviation = 0;
        
        for (int i = 1; i < track.detections.size() - 1; i++) {
            DeepSORTTracker.Detection prev = track.detections.get(i - 1);
            DeepSORTTracker.Detection curr = track.detections.get(i);
            DeepSORTTracker.Detection next = track.detections.get(i + 1);
            
            float deviation = calculatePerpendicularDistance(curr, prev, next);
            
            if (i > 1 && Math.signum(deviation) != Math.signum(previousDeviation)) {
                crossings++;
            }
            
            previousDeviation = deviation;
        }
        
        return crossings / timeSpan; // Frequency in Hz
    }

    private float calculateConcentration(int cellCount) {
        // Simplified concentration calculation
        // In real implementation, this would depend on chamber volume, dilution factor, etc.
        float chamberVolume = 0.1f; // mL
        float dilutionFactor = 1.0f;
        
        return (cellCount * dilutionFactor) / chamberVolume / 1000000; // Convert to millions/mL
    }
}